Log errors in RestResponse regardless of error_trace parameter

cbuescher opened the pull request:
Currently, setting the "error_trace" request parameter to "true" leads to error stack traces being
included in the rest response, but will prevent logging them in the "rest.suppressed" logger in the
`RestResponse` class. 
In order to improve visibility of those errors in our logs, this PR changes the logging behavior be 
independent of the "error_trace" request parameter and only take the "rest.exception.stacktrace.skip"
parameter into account, which is only used internally.

Closes #100884

cbuescher committed:
Log errors in RestResponse regardless of error_trace parameter

Currently we only log exceptions in RestResponse when the request parameter
"error_trace" isn't set or set to "false". If the client requests traces in the
rest response, we skip the server side logging. In order to improve visibility
of those errors in our logs this changes the logging behaviour to only depend on
the "rest.exception.stacktrace.skip" which is only used internally.

Closes #100884

elasticsearchmachine commented:
Hi @cbuescher, I've created a changelog YAML for you.

cbuescher committed:
Update docs/changelog/101066.yaml

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

cbuescher commented:
fyi I left the "rest.exception.stacktrace.skip" condition in there around the logging block because I'm unsure how we internally use it today. Some research makes me think that we only ever change the global default of "rest.exception.stacktrace.skip == true" in overwrites to the XContent rendering parameter for certain object, e.g. like in https://github.com/elastic/elasticsearch/blob/1e9c7f1d95598ed2be4a06953340f49d2fe22887/server/src/main/java/org/elasticsearch/action/admin/cluster/migration/GetFeatureUpgradeStatusResponse.java#L299. If that understanding is correct we might as well remove the logic here and in tests as well and only keep the parameter around for xcontent rendering purposes.

pgomulka reviewed (approved):
LGTM

javanna reviewed (commented):
Left one question, an additional one if we go ahead with this change (and I think we should) is whether the logger should still be called rest.suppressed. it seems like we are changing its behaviour significantly that it would deserve a rename as a follow-up, or at least a general discussion about the long-term direction.

quux00 commented on server/src/main/java/org/elasticsearch/rest/RestResponse.java:
```diff
@@ -131,6 +131,9 @@ public RestResponse(RestChannel channel, RestStatus status, Exception e) throws
                 SUPPRESSED_ERROR_LOGGER.warn(messageSupplier, e);
             }
         }
+        if (params.paramAsBoolean("error_trace", false) && status != RestStatus.UNAUTHORIZED) {
```
I know you are just rearranging code for this case, but (if you understand it) could you add a code comment of what this is for and what it is doing? It's not clear to me.

cbuescher replied:
I added comments

quux00 commented on server/src/main/java/org/elasticsearch/rest/RestResponse.java:
```diff
@@ -115,7 +115,7 @@ public RestResponse(RestChannel channel, Exception e) throws IOException {
         this.status = status;
-        ToXContent.Params params = paramsFromRequest(channel.request());
+        ToXContent.Params params = channel.request();
         if (params.paramAsBoolean(REST_EXCEPTION_SKIP_STACK_TRACE, REST_EXCEPTION_SKIP_STACK_TRACE_DEFAULT) && e != null) {
```
Do we need still need this check here?

```
params.paramAsBoolean(REST_EXCEPTION_SKIP_STACK_TRACE, REST_EXCEPTION_SKIP_STACK_TRACE_DEFAULT) 
```

Do we want to check if only if `status.getStatus() < 500` ?

cbuescher replied:
Reading this again more closely, REST_EXCEPTION_SKIP_STACK_TRACE is only ever set to "false" in GetFeatureUpgradeStatusResponse, ExceptionWatchRecord and ILMHistoryItem because it seems someone decided to always render exception stack traces inside those objects. That also means it should never be set here, so you are right this check can be removed. The presence of an exception is still something we'd need to check I think, also we want to log both 400/500s here so I don't see why we should condition on status here. I will push some changes.

cbuescher committed:
Remove superfluous condition around logging site and add comments

cbuescher commented:
@javanna @quux00 thanks for the additional review, I removed conditioning on the REST_EXCEPTION_SKIP_STACK_TRACE parameter in the RestResponse logging now since it cannot be different from the default value at least in this location. Its still used (and changed) for xcontent rendering so I'd like to keep it there for now. If we decide so we can also remove it there in a follow up. Let me know what you think.

quux00 reviewed (approved):
LGTM

cbuescher committed:
Update 101066.yaml
